proof/invariant-abstract/ARM/ArchTcb_AI.thy

```diff
@@ -210,6 +209,16 @@ lemma as_user_ipc_tcb_cap_valid4[wp]:
   apply (clarsimp simp: get_tcb_def)
   done
 
+(* Pull invs out of a complex goal and prove it only once. Use as (strengthen invs_strengthen)+,
+   best in combination with simp and potentially conj_cong *)
+lemma invs_strengthen:
+  "invs s \<Longrightarrow> P \<longrightarrow> invs s"
+  "invs s \<and> Q s \<Longrightarrow> P \<longrightarrow> invs s \<and> Q s"
+  "invs s \<and> Q s \<Longrightarrow> P \<longrightarrow> Q s \<and> invs s"
+  "invs s \<and> Q s \<Longrightarrow> P \<longrightarrow> (invs and Q) s"
+  "invs s \<and> Q s \<Longrightarrow> P \<longrightarrow> (Q and invs) s"
```

Comment:
I just realised that this does not only pull out `invs`, but also `Q`. Seems to be working just fine in the long-running monster tactic below (making it faster), but I was hoping that this might be more generally useful. I might have another look at this if it is still faster when `Q` is not pulled out.

Reply:
Doesn't make a noticeable time difference, so I've made it more general and moved it to `Invariants_AI`